### PR TITLE
Feature/created after

### PR DESCRIPTION
**Pull request by pond**
Support for default _opt-out_ framework-wide search/filter strings of `created_after` (for `created_at` timestamps after a particular given date/time) and `created_before` (self explanatory!).

Addresses #109.

One or two minor drive-by fixes included, such as removing the `dry_proc` in the search core code which wasn't really DRYing things much and obfuscated the code's operation quite badly.

RDoc comments have been reviewed and checked against RDoc output but the actual generated docs are not included in this PR for clarity and will follow in the subsequent version number bump PR once this the main change set is reviewed, agreed and merged.

### CODE REVIEW

File: lib/hoodoo/active/active_record/search_helper.rb [hunk 1]
```diff
@@ -14,8 +14,9 @@ module ActiveRecord
     module Finder
 
       # Help build up Hash maps to pass into Hoodoo::ActiveRecord::Finder
-      # methods Hoodoo::ActiveRecord::Finder#search_with and
-      # Hoodoo::ActiveRecord::Finder#filter_with.
+      # methods Hoodoo::ActiveRecord::Finder::ClassMethods#search_with and
+      # Hoodoo::ActiveRecord::Finder::ClassMethods#filter_with. Used also
+      # by the default framework search scopes.
```
Comment (pond) [inline lib/hoodoo/active/active_record/search_helper.rb]: Lots of drive-by RDoc fixes in this file plus some entirely new query functions added later.

File: lib/hoodoo/client/paginated_enumeration.rb [hunk 2]
```diff
@@ -52,7 +52,7 @@ def enumerate_all
           if results.size > 0
 
             if results.platform_errors.has_errors?
-              raise "Hoodoo::Client:: PaginatedEnumeration#enumerate_all: Unexpected internal state combination of results set and results error indication"
+              raise 'Hoodoo::Client::PaginatedEnumeration#enumerate_all: Unexpected internal state combination of results set and results error indication'
```
Comment (pond) [inline lib/hoodoo/client/paginated_enumeration.rb]: Drive-by fix.

File: spec/client/paginated_enumeration_spec.rb [hunk 3]
```diff
@@ -1,37 +1,34 @@
 require 'securerandom'
 require 'spec_helper.rb'
 
-#
 # These tests define the following Services.
```
Comment (pond) [inline spec/client/paginated_enumeration_spec.rb]: Lots of comment and style drive-by fixes in this, with no actual code alterations.

File: spec/client/paginated_enumeration_spec.rb [hunk 4]
```diff
@@ -1,37 +1,34 @@
...
-# │ ┌──────────────┐           ┌────────────────┐│ │  ┌───────────────────┐   │
-# │ │              │  inter    │                ││ │  │                   │   │
-# │ │ RSpecNumber  │◀resource ─│RSpecEvenNumber ││ │  │  RSpecOddNumber   │   │
-# │ │              │  local    │                ││ │  │                   │   │
-# │ └──────────────┘           └────────────────┘│ │  └───────────────────┘   │
-# │         ▲                                    │ │            │             │
-# │         │                            inter   │ │            │             │
-# │         └───────────────────────────resource ┼─┼────────────┘             │
-# │                                      remote  │ │                          │
-# └──────────────────────────────────────────────┘ └──────────────────────────┘
```
Comment (pond) [inline spec/client/paginated_enumeration_spec.rb]: UTF-8 multibyte characters in the above played havoc with some formatters and byte-vs-character count problems arose in a text editor, which is depressing for 2016 but the simple fix is to make the diagram pure ASCII! That's the change below.

File: spec/client/paginated_enumeration_spec.rb [hunk 5]
```diff
@@ -273,6 +263,15 @@ class RSpecRemoteNumberService < Hoodoo::Services::Service
...
     end
 
+    it 'detects internal inconsistencies' do
+      items = resources.first[ :endpoint ].list()
+      items.platform_errors.add_error( 'platform.timeout' )
+
+      expect {
+        items.enumerate_all { | item | }
+      }.to raise_error( RuntimeError, 'Hoodoo::Client::PaginatedEnumeration#enumerate_all: Unexpected internal state combination of results set and results error indication' )
+    end
```
Comment (pond) [inline spec/client/paginated_enumeration_spec.rb]: Drive-by test coverage fix to get the pagination system up to 100%.

File: lib/hoodoo/services/middleware/middleware.rb [hunk 6]
```diff
@@ -175,6 +175,46 @@ class Middleware
       }
     } )
 
+    # A validation Proc for FRAMEWORK_QUERY_DATA - see that for details. This
+    # one ensures that the value is a valid ISO 8601 subset date/time string
+    # and evalutes to the parsed version of that string if so.
```
Comment (mjnguyenloyalty) [inline lib/hoodoo/services/middleware/middleware.rb]: evaluates

File: lib/hoodoo/active/active_record/support.rb [hunk 7]
```diff
@@ -35,6 +35,51 @@ module ActiveRecord
...
+        #
+        # Note that the #search_with / #filter_with DSL declaration in an
+        # appropriately extended model can be used to override the default
+        # values wired in below, because the defaults are established by
+        # design _before_ the model declarations are processed.
+        #
+        mapping = {
+          'created_after'        => Hoodoo::ActiveRecord::Finder::SearchHelper.cs_gt( :created_at ),
+          'created_on_or_before' => Hoodoo::ActiveRecord::Finder::SearchHelper.cs_lte( :created_at )
+        }
```
Comment (mjnguyenloyalty) [inline lib/hoodoo/active/active_record/support.rb]: you created methods for cs_gte and cs_lt, but you don't make the mappings for those available here. Any reason why not?

File: lib/hoodoo/services/middleware/middleware.rb [hunk 8]
```diff
@@ -175,6 +175,46 @@ class Middleware
...
+    # DateTime.
+    #
+    # *IMPORTANT* - if this list is changed, any database support modules -
+    # e.g. in Hoodoo::ActiveRecord::Support - will need any internal mapping
+    # of "framework query keys to module-appropriate query code" updating.
+    #
+    FRAMEWORK_QUERY_DATA = {
+      'created_after'        => FRAMEWORK_QUERY_VALUE_DATE_PROC,
+      'created_on_or_before' => FRAMEWORK_QUERY_VALUE_DATE_PROC,
+    }
```
Comment (mjnguyenloyalty) [inline lib/hoodoo/services/middleware/middleware.rb]: If you do decide to expose created_on_or_after, or created_before, this would need to be updated.

Comment (pond) [inline lib/hoodoo/services/middleware/middleware.rb]: I was going to reply then realised I'd actually forgotten how this was meant to work due to interruption...

Searching on `created_after` returns all records after that date. By definition then, _filtering_ on that same quantity is the same as _searching_ on `created_on_or_before`. I should in fact have provided `created_after` and `created_before`, both exclusive, because when used in a filter context they provide the `...on_or_before` companions.

File: lib/hoodoo/active/active_record/support.rb (hunk 7 above)
Comment (mjnguyenloyalty) [inline lib/hoodoo/active/active_record/support.rb]: gotcha 👍

**Comment by mjnguyenloyalty** [comment]
👍  from me

**Comment by aames** [comment]
Reviewed test coverage with @pond
